core/src/main/java/org/wso2/testgrid/infrastructure/InfrastructureCombinationsProvider.java

```diff
@@ -131,4 +136,44 @@ private void addSubpropertiesAsInfrastructureParameters(InfrastructureCombinatio
         }
     }
 
+    /**
+     * This class provides set of filtered infrastructure values.
+     * <p>
+     * The filtering is executed by reading exclude and include attributes in the @{@link TestgridYaml} file.
+     * If exclude attribute is set in @{@link TestgridYaml} and if those are exists in the current infrastructure
+     * value set, those will be eliminated.
+     * Further, if include attribute is set in @{@link TestgridYaml} and if those are exists in the current
+     * infrastructure value set, only those infrastructure will be consider when creating combinations.
+     *
+     * @param infrastructures entire infrastructure value set
+     * @param testgridYaml    object model of testgrid.yaml config file
+     * @return filtered infrastructure value set
+     */
+    private Set<InfrastructureValueSet> filterInfrastructures(Set<InfrastructureValueSet> infrastructures,
+            TestgridYaml testgridYaml) {
+
+        List<String> excludes = testgridYaml.getInfrastructureConfig().getExcludes();
+        List<String> includes = testgridYaml.getInfrastructureConfig().getIncludes();
+        Set<InfrastructureValueSet> selectedInfraValSet = ConcurrentHashMap.newKeySet();
+
+        if (excludes != null && !excludes.isEmpty()) {
+            infrastructures.forEach(infrastructureValueSet -> {
+                Set<InfrastructureParameter> selectedSet = infrastructureValueSet.getValues().stream()
```

Comment:
Does infrastructureValueSet.getValues() method always return a value, unless there will be a possibility to happen NPE

Reply:
This method is executed when Test Grid generates test plans. In order to generate test plans, we can execute the following command and InfrastructureValueSet is constructed by reading infrastructure_parameter table.  Therefore infrastructureValueSet.getValues() method always return a value.
- generate-test-plan -p <product_name> --file <testgrid_yaml_path> --workingDir 
<working_directory_path>